File: rotate-image/rotate-image.py
class Solution:
    def transpose(self, arr):
        for i in range(len(arr)):
            for j in range(len(arr)):
                if i<j:
                    arr[i][j], arr[j][i] = arr[j][i], arr[i][j]
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.transpose(matrix)
        for i in matrix:
            i.reverse()

File: 48-rotate-image/rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++)
            for (int j =  i; j < matrix[i].length; j++)
                swapMatrix(matrix, i, j, j, i);

        for (int i = 0; i < matrix.length; i++)
            reverse(matrix[i], 0, matrix[i].length - 1);
    }

    void swapMatrix(int[][] arr, int i1, int i2, int j1, int j2) {
        int temp = arr[i1][i2];
        arr[i1][i2] = arr[j1][j2];
        arr[j1][j2] = temp;
    }

    void reverse(int[] arr, int l, int r) {
        while (l < r) {
            swap(arr, l, r);
            l++;
            r--;
        }
    }

    void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}